storage/src/object_store/ecs_s3_client.hpp

```diff
@@ -0,0 +1,412 @@
+// Copyright 2019 VMware, all rights reserved.
+
+#pragma once
+#include <libs3.h>
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <mutex>
+#include <thread>
+#include <functional>
+#include "Logger.hpp"
+#include "object_store/object_store_client.hpp"
+
+using namespace concord::storage;
+using namespace std;
+
+/**
+ * @brief Internal implementation of ECS EMC S3 client based on libs3
+ * Most of the method are irrelevant for the object store and throw exception if
+ * called. We assume that only ObjectStoreAppState class will use this class
+ *
+ */
+class EcsS3ClientImpl : public IDBClient {
+ public:
+  EcsS3ClientImpl(const S3_StoreConfig& config, const ObjectStoreBehavior& b) : config_{config}, bh_{b} {
+    LOG_INFO(logger_, "ECS S3 client created");
+  }
+
+  ~EcsS3ClientImpl() {
+    /* Destroy LibS3 */
+    S3_deinitialize();
+    init_ = false;
+    LOG_INFO(logger_, "libs3 deinit");
+  }
+
+  /**
+   * @brief used only from the test! name should be immutable string literal
+   *
+   * @param name
+   */
+  void set_bucket_name(string name) { config_.bucketName = name; }
+
+  /**
+   * @brief Initializing underlying libs3. The S3_initialize function must be
+   * called exactly once and only from 1 thread
+   *
+   * @param readOnly
+   */
+  virtual void init(bool readOnly) override {
+    std::lock_guard<std::mutex> g(initLock_);
```

Comment:
If this can be created and initialized from multiple threads, shouldn't some of the internal state also be protected? When would multiple threads try to initialize an ObjectStoreClient but not use it? Or is this because the rest of the functionality is already thread safe in the libS3 library?